### ISSUE TITLE
Calling seeders in beforeAll

### ISSUE THREAD

**Issue by mokhosh** [pestphp/pest#1094]
### What Happened

Imagine if you have a large seeder that takes 5s to run, and you need to test 20 things after that seed.

If you seed at the beginning of each test you would be waiting  a 100 seconds for those tests, but if you could run the seeder once, it will take about 5 seconds, which is a massive improvement.

Please provide a way for us to do this:
```php
beforeAll(fn () => seed());
```

This is also mentioned in #237 and #246 and https://github.com/pestphp/pest/issues/33#issuecomment-658022624
but sadly they are closed and ignored, but this seems to be a big pain point and a deal breaker for a lot of people.

### How to Reproduce

Do `beforeAll(fn () => seed());` in a pest test file and you will get this:
```
  assert(array_key_exists(self::FILE, $trace))

  at tests/Unit/SomeTest.php:7
      3▕ use function Pest\Laravel\assertDatabaseHas;
      4▕ use function Pest\Laravel\seed;
      5▕ 
      6▕ beforeAll(function () {
  ➜   7▕     seed();
      8▕ });

```

### Pest Version

2.33.66

### PHP Version

8.3.2

### Operation System

macOS

**Comment by GolearnMkieler** [comment]
Still no fix for this i guess.. Why is this ignored?!

**Comment by djurovicigoor** [comment]
Need this...

**Comment by djurovicigoor** [comment]
@nunomaduro

**Comment by zealouscreations** [comment]
@nunomaduro This is a show-stopper. If I can't use `beforeAll` and `afterAll` like PHPUnit `setUp` and `tearDown` where I'd still have access to the initialized Laravel app, then this renders Pest unusable for me and I'm switching back to PHPUnit.

**Comment by xvargr** [comment]
This is such an obvious use case, I'm shocked it isn't supported

**Comment by GringoDotDev** [comment]
In case it's helpful for anyone else, Pest is built on PHPUnit and you can use PHPUnit features with Pest pretty easily. Just look for a line like this in your Pest.php:

`uses(TestCase::class)->in("Feature");`

TestCase is a normal PHPUnit class that you can customize with whatever hooks you want. I'm actually doing some crazy stuff in mine like bootstrapping tenancy, etc. Seeders are possible in setUpBeforeClass() with the normal workarounds (calling the command via artisan, etc.).

You can have as many parent classes as you want divided up however you want.

Separately, I do feel like some of these comments come across as a bit "latosos" (as we say in Mexico) so I'd  just like to take the opportunity to thank @nunomaduro for building such an awesome test framework and giving it to the world for free.

**Comment by abbasali** [comment]
Maybe I'm not fully understanding the use case, but can't we simply add the following to the base `TestCase` class?

```php
abstract class TestCase extends BaseTestCase
{
    public bool $seed = true;
}
```

and then use `RefreshDatabase` trait in `Pest.php`

```php
use Illuminate\Foundation\Testing\RefreshDatabase;

pest()->extend(Tests\TestCase::class)
    ->use(RefreshDatabase::class)
    ->in('Feature');
```

and this will seed the database only once at the start of the testsuite.

The `Illuminate\Foundation\Testing\Traits\CanConfigureMigrationCommands` trait (used by `RefreshDatabase`) already contains the logic to run the seeder if `$seed` is set to true, or we can provide a custom seeder class with the `$seeder` property. The seeding is done only once when the `migrate:fresh` command is executed in the `RefreshDatabase::refreshTestDatabase()` method.

**Comment by mokhosh** [comment]
@abbasali @GringoDotDev 

These are good suggestions if you want to run something before all tests of an entire suit.

But with beforeAll you can have much more control and for example do something before all tests of one file.

**Comment by abbasali** [comment]
Yes, what I suggested is only useful if you want to run the seeder only once for the entire TestSuite. There is no easy/direct way of doing it for each file at the moment. A workaround may be to set a `$seeded` static property on the test class with initial value of `false` and then in `beforeEach()` seed the database if that property is `false` and set it to `true` once seeded. That way the seeder will run only once for the entire file. Haven't tried it but something like this may work.

**Comment by shavonn** [comment]
> Yes, what I suggested is only useful if you want to run the seeder only once for the entire TestSuite. There is no easy/direct way of doing it for each file at the moment. A workaround may be to set a `$seeded` static property on the test class with initial value of `false` and then in `beforeEach()` seed the database if that property is `false` and set it to `true` once seeded. That way the seeder will run only once for the entire file. Haven't tried it but something like this may work.

I did something like that, and it worked, so I went with it.

```
$categoriesSeeded = false;

beforeEach(function () use (&$categoriesSeeded) {
    if (!$categoriesSeeded) {
        $this->seed(CategorySeeder::class);
    }
    $categoriesSeeded = true;
});
```

**Comment by mokhosh** [comment]
@shavonn you're not refreshing the db?

**Comment by GringoDotDev** [comment]
> @abbasali @GringoDotDev
> 
> These are good suggestions if you want to run something before all tests of an entire suit.
> 
> But with beforeAll you can have much more control and for example do something before all tests of one file.

Totally. Being able to do it in beforeAll would be super convenient. But this technique doesn't necessarily have to apply to the entire test suite. Different test files can inherit from different base PHPUnit TestCase classes. One of those base classes could have a seeder, and one of them could be vanilla.

**Comment by haythembenkhlifa** [comment]
thank you @abbasali

**Comment by kavanbsf** [comment]
@nunomaduro

**Comment by kavanbsf** [comment]
If you want to prepare data or execute code before running the test suite, you can use this simple technique:

`Custom Technique:`
Create a test file named `AA00.php` inside `tests/Feature`. The alphabetical naming ensures this file runs first, before any other test executes.

You can write any global setup logic inside this file, such as seeding data, initializing configurations, or running necessary commands.

**Comment by clemblanco** [comment]
It feels like being able to call something like `pest()->beforeEverything()` and `pest()->afterEverything()` from `tests/Pest.php` is kinda missing... It would be executing a hook once per test suite run.

I believe Jest uses [globalSetup](https://jestjs.io/docs/configuration#globalsetup-string) and [globalTeardown](https://jestjs.io/docs/configuration#globalteardown-string) for this.

> The function will be triggered once before/after all test suites...

Feels like a good opportunity for a contribution here.

In the meantime, I think we should be able to manually rely on the underlying functionality from Laravel/PHPUnit by using your own `TestCase.php` and define some hooks using `afterApplicationCreated()` and `beforeApplicationDestroyed()` coming from `Illuminate\Foundation\Testing\Concerns\InteractsWithTestCaseLifecycle`... This is how `RefreshDatabase` works.

**Comment by clemblanco** [comment]
Technically with Laravel, you can rely on this https://laravel-news.com/laravel-test-case-trait-setup-hooks and have:

```php
namesapce Tests\Concerns;

trait FooBar
{
    public function setUpFooBar()
    {
        // Executed once when setting up the entire test suite...
    }

    public function tearDownFooBar()
    {
        // Executed once when tearing down the entire test suite...
    }
}
```
and use it with
```php
pest()->extend(Tests\TestCase::class)
    ->use(Tests\Concerns\FooBar::class)
```
and in theory it should work.

If you're using Pest without Laravel, I'm not sure Pest has got something out of the box (yet).

**Comment by nunomaduro** [comment]
i understand why you guys want this; and i also understand that probably there is a better way to solve this. however, today, the beforeAll is behaving exactly like the before all does in phpunit. so, whatever we do here, will be a new feature, not a bug fix.
